gmso/core/topology.py

```diff
@@ -543,48 +543,88 @@ def update_connection_types(self):
-                    c.connection_type = self._bond_types[c.connection_type]
-                if isinstance(c.connection_type, AngleType):
-                    c.connection_type = self._angle_types[c.connection_type]
-                if isinstance(c.connection_type, DihedralType):
-                    c.connection_type = self._dihedral_types[c.connection_type]
-                if isinstance(c.connection_type, ImproperType):
-                    c.connection_type = self._improper_types[c.connection_type]
+            if not isinstance(connection_type_or_types, IndexedSet):
+                connection_type_or_types = [connection_type_or_types]
+            for connection_type in connection_type_or_types:
+                if connection_type not in self._connection_types:
+                    connection_type.topology = self
+                    self._connection_types[connection_type] = connection_type
+                    if isinstance(connection_type, BondType):
+                        self._bond_types[connection_type] = connection_type
+                        self._bond_types_idx[connection_type] = (
+                            len(self._bond_types) - 1
+                        )
+                    if isinstance(connection_type, AngleType):
+                        self._angle_types[connection_type] = connection_type
+                        self._angle_types_idx[connection_type] = (
+                            len(self._angle_types) - 1
+                        )
+                    if isinstance(connection_type, DihedralType):
+                        self._dihedral_types[connection_type] = connection_type
+                        self._dihedral_types_idx[connection_type] = (
+                            len(self._dihedral_types) - 1
+                        )
+                    if isinstance(connection_type, ImproperType):
+                        self._improper_types[
+                            connection_type
+                        ] = c.connection_type
+                        self._improper_types_idx[connection_type] = (
+                            len(self._improper_types) - 1
+                        )
+                elif connection_type in self.connection_types:
+                    if isinstance(connection_type, BondType):
+                        c.connection_type = self._bond_types[connection_type]
+                    if isinstance(connection_type, AngleType):
+                        c.connection_type = self._angle_types[connection_type]
+                    if isinstance(connection_type, DihedralType):
+                        if c.is_layered():
+                            if c.connection_types is None:
+                                c.connection_types = IndexedSet(
+                                    [self._dihedrals_types[connection_type]]
+                                )
+                            else:
+                                c.connection_types.add(
```

Comment:
Should add a comment about this.

Based on my understanding of IndexedSet, if an object is attempted to be added to the indexedSet, and it generates a hash equivalent to a member already in the set, then the object that is trying to be added to the set will be ignored.

```python3
a_hash = hash(a) # 100
b_hash = hash(b) # 100

iset = IndexedSet(a)
iset.add(b) # should not be added to the indexed set
len(iset) # 1
```